File: app/src/main/java/com/mbg/mbgsupport/demo/java/singleton/DemoSingletonFourth.java
package com.mbg.mbgsupport.demo.java.singleton;

/***
 * java方式双重认证锁创建单例的demo ---推荐使用方式
 * created by canwen2003
 */
public class DemoSingletonFourth {
    private static final String TAG="DemoSingletonFourth";
    private static DemoSingletonFourth instance;
    private DemoSingletonFourth(){

    }
    public static DemoSingletonFourth getInstance(){
        if (instance==null){
            synchronized (DemoSingletonFourth.class) {
                if (instance==null){
                    instance = new DemoSingletonFourth();
                }
            }
        }
        return instance;
    }

    public void logInfo(String str){
        System.out.println(TAG+":"+str);
    }
}

File: app/src/main/java/com/mbg/mbgsupport/demo/kotlin/singleton/DemoSingletonFourth.kt
package com.mbg.mbgsupport.demo.kotlin.singleton

/***
 * Kotlin方式双重认证锁创建单例的demo ---推荐使用方式
 * created by canwen2003
 */
class DemoSingletonFourth private constructor(){
    fun logInfo(str: String){
        println("${TAG}:$str")
    }

    companion object{
        private const val TAG="DemoSingletonFourth"
        private val instance: DemoSingletonFourth by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            DemoSingletonFourth()
        }

        fun get():DemoSingletonFourth{
            return instance;
        }
    }
}